test/distributed/fsdp/test_fsdp_meta.py

```diff
@@ -0,0 +1,315 @@
+        def meta_module_fn():
+            return MyModel(device="meta")
+        self._test_simple_model_with_meta_device(
+            meta_module_fn, _init_with_reset_params
+        )
+
+    @skip_if_lt_x_gpu(2)
+    def test_simple_model_with_meta_device_default_init(self):
+        def meta_module_fn():
+            return MyModel(device="meta")
+        self._test_simple_model_with_meta_device(meta_module_fn)
+
+    @skip_if_lt_x_gpu(2)
+    @sandcastle_skip_if(
+        not _TORCHDISTX_AVAIL, "Test requires torchdistx: https://github.com/pytorch/torchdistx"
+    )
+    def test_simple_model_with_torchdistx_default_init(self):
+        def meta_module_fn():
+            return deferred_init.deferred_init(MyModel, device="cuda")
+
+        self._test_simple_model_with_meta_device(meta_module_fn)
+
+    @skip_if_lt_x_gpu(2)
+    @sandcastle_skip_if(
+        not _TORCHDISTX_AVAIL, "Test requires torchdistx: https://github.com/pytorch/torchdistx"
+    )
+    def test_simple_model_with_torchdistx_init_fn(self):
+        def meta_module_fn():
+            return deferred_init.deferred_init(MyModel, device="cuda")
+
+        self._test_simple_model_with_meta_device(meta_module_fn, init_fn=_init_with_torchdistx)
+
+    def _test_nested_model_with_meta_device(self, auto_wrap, meta_module_fn, init_fn=None):
+        if auto_wrap:
+            module = meta_module_fn()
+            is_meta = next(module.parameters()).is_meta
+            fsdp_meta = FSDP(
+                module,
+                auto_wrap_policy=always_wrap,
+                param_init_fn=init_fn,
+            )
+            meta_opt = torch.optim.SGD(fsdp_meta.parameters(), lr=1e-3)
+            module_regular = NestedModel(device="cuda")
+            # For torchdistx init, we don't need to call reset_params, as
+            # deferred_init(model).materialize() is equivalent to model().
+            if is_meta:
+                module_regular.reset_parameters()
+            fsdp_regular = FSDP(
+                module_regular,
+                auto_wrap_policy=always_wrap,
+            )
+            regular_opt = torch.optim.SGD(fsdp_regular.parameters(), lr=1e-3)
+        else:
+            with enable_wrap(
+                wrapper_cls=FSDP, param_init_fn=init_fn,
+            ):
+                module = meta_module_fn()
+                is_meta = next(module.parameters()).is_meta
+                # Non FSDP modules will still be initialized because they bubble up
+                # to be part of a larger FSDP unit.
+                fsdp_meta = wrap(module)
+                meta_opt = torch.optim.SGD(fsdp_meta.parameters(), lr=1e-3)
+
+            # Init and reset parameters before wrapping so that reset_params
+            # matches up with meta device's initialization.
+            module_regular = NestedModel(device="cuda")
+            # For torchdistx init, we don't need to call reset_params, as
+            # deferred_init(model).materialize() is equivalent to model().
```

Comment:
Maybe just refactor this small piece of logic into a method and document the explanation once there since it is showing up multiple times.

```
def <choose a name>(self, is_meta: bool, module: torch.nn.Module):
    # Add explanation here
    if is_meta:
        module.reset_parameters()
```

Reply:
refactored into common code